src/BrandwatchReactAuth.js

```diff
@@ -29,13 +29,22 @@ export default class BrandwatchReactAuth extends Component {
   }
 
   componentWillMount() {
-    this.handleGetToken().then((token) => {
-      if (!token) {
-        return window.location.replace(this.store.loginUrl);
-      }
+    const { backupDomain, backupRedirect } = this.props;
 
-      this.setState(() => ({ loggedIn: true }));
-    });
+    this.handleGetToken()
+      .then(token => {
+        return token
+          ? this.setState(() => ({ loggedIn: true }))
+          : new Promise((resolve, reject) => {
+            return backupDomain ? resolve(this.handleGetToken(backupDomain)) : reject()
+          })
+          .then(token => {
+            return token && backupRedirect ? window.location.replace(backupRedirect) : Promise.reject()
+          })
+          .catch(e => {
+            return window.location.replace(this.store.loginUrl)
+          })
+      });
```

Comment:
Definitely a challenging flow to write. I do find it a bit difficult to follow with all those ternaries and several formats of creating promises. I'm all for the short hand syntax, but I wonder if this complicated flow would be easier to read with something a bit more step based. What do you think about something along these lines... 

```js
this.handleGetToken().then((token) => {
  if (token) {
    this.setState(() => ({ loggedIn: true }));
  } else if (backupDomain) {
    this.handleGetToken(backupDomain).then((token) => {
      if (token && backupRedirect) {
        window.location.replace(backupRedirect);
      }
    });
  } else {
    window.location.replace(this.store.loginUrl);
  }
});
```

Reply:
I think I agree with you, but when I ran it through the tests it turns out it would need to be this...

```js
    this.handleGetToken().then((token) => {
      if (token) {
        this.setState(() => ({ loggedIn: true }));
      } else if (backupDomain) {
        console.log('if backupDomain');
        this.handleGetToken(backupDomain).then((token) => {
          console.log('crash');
          if (token && backupRedirect) {
            window.location.replace(backupRedirect);
          } else {
            window.location.replace(this.store.loginUrl);            
          }
        });
      } else {
        window.location.replace(this.store.loginUrl);
      }
    });
```